lib/ansible/modules/identity/keycloak/keycloak_realm.py

```diff
@@ -0,0 +1,1063 @@
+                           supports_check_mode=True,
+                           required_one_of=([['name', 'realm']]))
+
+    result = dict(changed=False, msg='', diff={}, proposed={}, existing={}, end_state={})
+
+    # Obtain access token, initialize API
+    kc = KeycloakAPI(module)
+
+    state = module.params.get('state')
+    realm = module.params.get('name') if module.params.get('name') is not None else module.params.get('realm')
+
+    # convert module parameters to realm representation parameters (if they belong in there)
+    realm_params = [x for x in module.params
+                    if x not in list(keycloak_argument_spec().keys()) + ['state', 'name'] and
+                    module.params.get(x) is not None]
+
+    # If the I(realm) is not set, assume I(name) as the default.
+    if module.params.get('realm') is None:
+        realm_params['realm'] = realm
+
+    before_realm = kc.get_realm_by_name(realm)
+    before_realm = {} if before_realm is None else before_realm
+
+    # Build a proposed changeset from parameters given to this module
+    changeset = dict()
+
+    for realm_param in realm_params:
+        # lists in the Keycloak API are sorted
+        new_param_value = module.params.get(realm_param)
+        if isinstance(new_param_value, list):
+            try:
+                new_param_value = sorted(new_param_value)
+            except TypeError:
+                pass
+        changeset[camel(realm_param)] = new_param_value
+
+    # Whether creating or updating a realm, take the before-state and merge the changeset into it
+    updated_realm = before_realm.copy()
+    updated_realm.update(changeset)
+
+    result['proposed'] = changeset
+    result['existing'] = before_realm
+
+    # If the realm does not exist yet, before_realm is still empty
+    if before_realm == dict():
+        if state == 'absent':
+            # do nothing and exit
+            if module._diff:
+                result['diff'] = dict(before='', after='')
+            result['msg'] = 'Realm does not exist, doing nothing.'
+            module.exit_json(**result)
+
+        # create new realm
+        result['changed'] = True
+
+        # when creating a realm, the id should be the same as the realm name -- otherwise bugs magically appear
+        if 'id' not in updated_realm:
+            updated_realm['id'] = updated_realm['realm']
+
+        if module.check_mode:
+            if module._diff:
+                result['diff'] = dict(before='', after=updated_realm)
+
+            module.exit_json(**result)
+
+        kc.create_realm(updated_realm)
+        after_realm = kc.get_realm_by_name(realm)
+
+        if module._diff:
+            result['diff'] = dict(before='', after=after_realm)
+
+        result['end_state'] = after_realm
+
+        result['msg'] = 'Realm %s has been created.' % realm
+        module.exit_json(**result)
+    else:
+        if state == 'present':
+            # update existing realm
+            result['changed'] = True
```

Comment:
Rather than setting to True, can you do what's happening for check mode `result['changed'] = (before_realm != updated_realm)` and check it's True before making the API call to update? Or is that not necessarily accurate for check mode? I'm not sure what the format of before_realm versus updated_realm is.

Reply:
It's not necessarily accurate. The check-mode version is a bit dumber than the API endpoint (the latter has some agnosticity regarding datatypes for instance). While improving check mode would definitely be good, I don't think I can always model what the API will do with any given input, so instead the changed gets set to true by default and later changed to false if the API returns the same data after the update call as it did before.

before_realm is just the API call result before anything has changed, updated_realm has the proposed changeset applied to it.